course/management/commands/delete_instance.py

```diff
@@ -0,0 +1,53 @@
+from django.core.management.base import BaseCommand
+from course.models import CourseInstance
+import datetime
+
+class Command(BaseCommand):
+    help = "Delete course instance along with related data (exercises, submissions, etc)."
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-c',
+            '--course-instance',
+            type=int,
+            help='Course instance ID to be deleted.',
+        )
+        parser.add_argument(
+            '-e',
+            '--ending-since',
+            type=int,
+            help='Delete all course instances that ended since N days ago.',
+        )
+        parser.add_argument(
+            '--dry-run',
+            action='store_true',
+            help='Do not delete anything, but show which instances would be deleted.',
+        )
+
+    def handle(self, *args, **options):
+        course_instances: list[CourseInstance] = []
```

Comment:
Is it necessary to use a list? Iterating over a QuerySet is more efficient than storing all course instance objects in memory in a list at once. QuerySets with different conditions could be combined with `.union()`.
https://docs.djangoproject.com/en/3.2/ref/models/querysets/#union

Reply:
fixed